src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/Debugging/RemoteDebugInfoService.cs

```diff
@@ -0,0 +1,156 @@
+        var sourceText = await context.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+        var hostDocumentIndex = sourceText.GetPosition(position);
+
+        if (!TryGetUsableProjectedIndex(codeDocument, hostDocumentIndex, out var projectedIndex))
+        {
+            return null;
+        }
+
+        // Now ask Roslyn to adjust the breakpoint to a valid location in the code
+        var generatedDocument = await context.Snapshot.GetGeneratedDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var syntaxTree = await generatedDocument.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+        if (syntaxTree is null)
+        {
+            return null;
+        }
+
+        if (!RazorBreakpointSpans.TryGetBreakpointSpan(syntaxTree, projectedIndex, cancellationToken, out var csharpBreakpointSpan))
+        {
+            return null;
+        }
+
+        var csharpText = codeDocument.GetCSharpSourceText();
+        var projectedRange = csharpText.GetLinePositionSpan(csharpBreakpointSpan);
+
+        // Inclusive mapping means we are lenient to portions of the breakpoint that might be outside of use code in the Razor file
+        if (!_documentMappingService.TryMapToHostDocumentRange(codeDocument.GetCSharpDocument(), projectedRange, MappingBehavior.Inclusive, out var hostDocumentRange))
+        {
+            return null;
+        }
+
+        cancellationToken.ThrowIfCancellationRequested();
+
+        return hostDocumentRange;
+    }
+
+
+    public ValueTask<string[]?> ResolveProximityExpressionsAsync(RazorPinnedSolutionInfoWrapper solutionInfo, DocumentId documentId, LinePosition position, CancellationToken cancellationToken)
+        => RunServiceAsync(
+            solutionInfo,
+            documentId,
+            context => ResolveProximityExpressionsAsync(context, position, cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<string[]?> ResolveProximityExpressionsAsync(RemoteDocumentContext context, LinePosition position, CancellationToken cancellationToken)
+    {
+        var codeDocument = await context.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var sourceText = await context.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+        var hostDocumentIndex = sourceText.GetPosition(position);
+
+        if (!TryGetUsableProjectedIndex(codeDocument, hostDocumentIndex, out var projectedIndex))
+        {
+            return null;
+        }
+
+        // Now ask Roslyn to adjust the breakpoint to a valid location in the code
+        var generatedDocument = await context.Snapshot.GetGeneratedDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var syntaxTree = await generatedDocument.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+        if (syntaxTree is null)
+        {
+            return null;
+        }
```

Comment:
Is this first part exactly the same as the previous method, ResolveBreakpointRangeAsync?

Reply:
Yes

Reply:
helper method for the common part? :)

Reply:
IMO sharing it all is a little awkward, but I found an existing extension method to get a syntax tree, and moved a little bit of code into the helper method, so its simpler now. Thanks